File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        minval = prices[0]
        for i in range(1, len(prices)):
            profit = prices[i] - minval
            res = max(profit, res)
            minval = min(prices[i], minval)
        return res

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0];
        int curMax = 1, curMin = 1;
        for(int num : nums) {
            int temp = curMax;
            curMax = max(num, max(temp * num, curMin * num));
            curMin = min(num, min(temp * num, curMin * num));
            res = max(res , curMax);
        }
        return res;
    }
};
